shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -287,6 +287,10 @@ private static int clampIndexToEditable(int index, Editable editable) {
   // occur, and need a chance to be handled by the framework.
   @Override
   public boolean sendKeyEvent(KeyEvent event) {
+    if (event.getAction() == KeyEvent.ACTION_DOWN
```

Comment:
Sorry I have not looked into this. Do you mind briefly explaining what caused the brackets to be replaced by the old input?

Reply:
The [issue](https://github.com/flutter/flutter/issues/120351) has a link to a doc in the comments where I tried to capture what is happening (reposted [here](https://docs.google.com/document/d/1wXqJ5kc_PmxK5eTeFZA3C0dKa6vbIAmkQuSTKqEJx5s/edit#)). As far as I can tell, The Samsung IME is sending an additional key left event that is not being handled properly (it's likely getting swallowed somewhere in the KeyboardDelegate - this is where I'm getting lost and could use some guidance) which results in the brackets not getting committed because they are being invalidated due to the cursor location not matching what is expected. So, the originally typed text that should be replaced by the brackets is reverting instead.

Reply:
AFAIK the usual procedure for handling key events is to first deliver the key event to the framework, and let the framework decides what to do with it. If the framework does not handle the key event, the key event gets sent back to the embedder. This allows developers to define shortcuts in the framework. Applying this patch would likely disable that for `<-` alone. When `<-` is pressed and released the framework would receive only the up event without the down event so that could be problematic.

Reply:
@dkwingsmt are the `KEYCODE_DPAD_*` keys in our keyboard maps for Android?

Reply:
Summary of discussion over chat.

Based on the videos in the bug, it looks like immediately after the autocompletion, the engine probably has the right state and sends it to the framework (we see the かっこ gets replaced with brackets) but that sometime after the engine sends the text editing update that replaces the かっこ text with [], but before the framework applies it, somehow the engine is getting reset to the old かっこ　state. The right fix is probably to figure out what's triggering that state reset and prevent it, so that we avoid suppressing key events to the framework.

I _suspect_ that as-is, this will likely break apps that want to handle arrow key events for non text editing scenarios -- e.g. games, keyboard shortcuts, etc. which would impact tablets and chromebooks running android apps with keyboards (assuming it's the same event?), if not then maybe game controllers with a d-pad. 

If the issue turns out to be that we're handling the dpad-left key event in the framework before we get the updated text (I'd expect it to come after the updated text), then I think the question is _why_ we're handling that event in the framework at all rather than allowing Android's text input system to handle it, just like it does other text editing events like pressing the 'a' key, for example. Otherwise, I'm curious what's triggering the state to get reset to the old state in the engine - my guess is an update from the framework, but triggered by what?